src/Intl/Transliterator/Transliterator.php

```diff
@@ -0,0 +1,761 @@
+    }
+
+    public static function createFromRules($rules, $direction = \Transliterator::FORWARD)
+    {
+        if (
+            false === strpos($rules, '<')
+            &&
+            false === strpos($rules, '>')
+            &&
+            false === strpos($rules, '::')
+        ) {
+            return null;
+        }
+
+        $transliterator = new self();
+
+        $transliterator->id_intern = self::clean_id($rules);
+
+        if (\Transliterator::FORWARD !== $direction && null !== $direction) {
+            throw new \DomainException(sprintf('The PHP intl extension is required for using "Transliterator->direction".'));
+        }
+
+        return $transliterator;
+    }
+
+    public static function createInverse()
+    {
+        throw new \DomainException(sprintf('The PHP intl extension is required for using "Transliterator::createInverse".'));
+    }
+
+    /**
+     * Get data from "/data/*.php".
+     *
+     * @param string $file
+     *
+     * @return array
+     */
+    private static function getData($file)
+    {
+        /* @noinspection PhpIncludeInspection */
+        /* @noinspection UsingInclusionReturnValueInspection */
+        /* @psalm-suppress UnresolvableInclude */
+        return include __DIR__.'/Resources/unidata/'.$file.'.php';
+    }
+
+    /**
+     * Get data from "/data/*.php".
+     *
+     * @param string $file
+     *
+     * @return array|false
+     *                     <p>Will return <strong>false</strong> on error.</p>
+     */
+    private static function getDataIfExists($file)
+    {
+        $file = __DIR__.'/Resources/unidata/'.$file.'.php';
+        if (file_exists($file)) {
+            /* @noinspection PhpIncludeInspection */
+            /* @noinspection UsingInclusionReturnValueInspection */
+            return include $file;
+        }
+
+        return false;
+    }
+
+    public function getErrorCode()
+    {
+        return 0;
+    }
+
+    public function getErrorMessage()
+    {
+        return '';
+    }
+
+    /**
+     * Get the language from a string.
+     *
+     * e.g.: de_at -> de_at
+     *       de_DE -> de
+     *       DE_DE -> de
+     *       de-de -> de
+     *
+     * @param string $language
+     *
+     * @return string
+     */
+    private static function get_language($language)
+    {
+        if ('' === $language) {
+            return '';
+        }
+
+        if (
+            false === strpos($language, '_')
+            &&
+            false === strpos($language, '-')
+        ) {
+            return strtolower($language);
+        }
+
+        $regex = '/(?<first>[a-z]+)[\-_]\g{first}/i';
+
+        return str_replace(
+            '-',
+            '_',
+            strtolower(
+                (string) preg_replace($regex, '$1', $language)
+            )
+        );
+    }
+
+    public static function listIDs()
+    {
+        return array_values(self::$LOCALE_TO_TRANSLITERATOR_ID);
+    }
+
+    private function normalize_regex_helper($rule)
```

Comment:
using replacement in the rule to convert it to a regex has a big drawback: it could very easily allow injecting whather we want in the regex, breaking everything. I haven't tested this, but I'm almost sure that garbage is rules could go through (at least some of it)

Reply:
Jep, I have tried to make this more stable in my last commit. So now we only have one "preg_replace" + one "preg_replace_callback" call and we will check the regex via "@preg_match" before we use it. 

---

Or we only try to polyfill "create" and not "createFromRules" in the first place?

Reply:
my point is that you should probably not replace in the original string, while letting any unmatched part unchanged, but rather parse the original string and explicitly deal with *all* parts.

A single implementation of replacement does not solve the fact that you could inject stuff between placeholders being replaced.

Reply:
Can you please give me a example? I didn't 100% understand what you mean.

We already split the "rules" (by ";") into chunks, and then normalize the regex and we use only the part in brackets for the regex.

Reply:
@stof ping

Reply:
I tried, but the problem is that the syntax from the ICU's regex is not 100% compatible with the PCRE regex, so that it looks like I have to modify the original regex, so that PHP can run it. :-/ Any further idea how to solve this problem?

Reply:
@ThomasWeinert maybe you can help me with this problem? I try to write a polyfill for " Transliterator", but it use ICU regex. Is there a library that I can relay on or do you have any idea how to handle this?

Reply:
The clean way would be to implement a small parser for the transliteration rules and compile the PCRE pattern from the tokens/ast. It would allow to throw specific errors.

Reply:
ICU rules are not a regex, so that's expected that they are not compatible with PCRE. My whole point is that you should not be transforming the ICU rules to a PCRE regex (silently keeping anything that your transformation does not explicitly transforms) but you should parse the rules and generate a PCRE regex, so that nothing can be silently injected in the regex.

Reply:
@stof ICU rules are using the ICU regular expression syntax, (http://userguide.icu-project.org/strings/regexp) so I think we need an parser for that, otherwise I don't know how to generate a PCRE regex from a ICU regex. And because there are so many options, I don't think that I can simply parse the regex stuff.

@ThomasWeinert Thanks for the feedback. Do you think I can use something like https://github.com/hoaproject/Regex to parse the ICU regex grammar https://github.com/gagolews/stringi/blob/master/src/icu61/i18n/regexcst.txt ?

Reply:
I used only transliteration **identifiers** before: http://userguide.icu-project.org/transforms/general#TOC-Formal-ID-Syntax

Transliteration **rules** would be for designing your own transliterations: http://userguide.icu-project.org/transforms/general#TOC-Designing-Transliterators 

*btw* http://userguide.icu-project.org/transforms/general#TOC-Regular-Expression

Reply:
AFAICS this is hopeless. The [Transform Rules Syntax of UTS #35: Unicode LDML](http://www.unicode.org/reports/tr35/tr35-39/tr35-general.html#Transform_Rules_Syntax) is easy to understand for a human (and I am glad it is so because the ICU docs are anything but) but I do not see a formal grammar.

As far as I can see the [reference source code](https://github.com/unicode-org/icu/blob/d3315d98ef82b09aabef12e3f7fb46d171d8bc32/icu4c/source/i18n/rbt_pars.cpp) does _not_ use the regex engine and I can see significant [differences](https://github.com/unicode-org/icu/blob/d3315d98ef82b09aabef12e3f7fb46d171d8bc32/icu4c/source/i18n/rbt_pars.cpp#L647) differences even in matching. And then there's variable definitions and function calls. All this makes me question whether anything else but porting this monstrosity wholesale would achieve full compatibility. In other words, here's the beginning of the Latin-ASCII ruleset 

`:: [[:Latin:][:Common:][:Inherited:][〇]] ;:: NFD() ;`

I think a discussion of what is feasible to support and then sharply limiting the rules to those is the way forward.